### PR TITLE
PERF: CategoricalDtype.__eq__

### PR DESCRIPTION
**Pull request by jbrockmendel**
Three special cases where we can get better performance

1) categories have different lengths --> never equal
2) categories have different dtypes --> never equal
3) categories are not object-dtype --> can use use get_indexer instead of hashing

```
import pandas as pd
import numpy as np

dti = pd.date_range("2016-01-01", periods=10000, tz="US/Pacific")
dti2 = type(dti)(dti._data.copy())

np.random.shuffle(dti._data._data)
assert not dti.equals(dti2)

cd1 = pd.CategoricalDtype(dti)
cd2 = pd.CategoricalDtype(dti2)
cd3 = pd.CategoricalDtype(dti2[:-1])
cd4 = pd.CategoricalDtype(pd.Index(range(len(dti))))

In [5]: %timeit cd1 == cd2 
548 µs ± 3.94 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- master
239 µs ± 1.04 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- PR

In [6]: %timeit cd1 == cd3
543 µs ± 4.05 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)   # <-- master
5.05 µs ± 257 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each) # <-- PR

In [7]: %timeit cd1 == cd4
389 µs ± 13.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <-- master
2.17 µs ± 63.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # <-- PR

```

### CODE REVIEW

File: pandas/core/dtypes/dtypes.py [hunk 1]
```diff
@@ -375,12 +375,28 @@ def __eq__(self, other: Any) -> bool:
...
+                return False
+
+            if len(left) != len(right):
+                return False
+
+            if self.categories.equals(other.categories):
                 # Check and see if they happen to be identical categories
                 return True
+
+            if left.dtype != object:
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:390]: use is_dtype_equal?

File: pandas/core/dtypes/dtypes.py [hunk 2]
```diff
@@ -375,12 +375,28 @@ def __eq__(self, other: Any) -> bool:
             # but same order is not necessary.  There is no distinction between
             # ordered=False and ordered=None: CDT(., False) and CDT(., None)
             # will be equal if they have the same categories.
-            if (
-                self.categories.dtype == other.categories.dtype
-                and self.categories.equals(other.categories)
-            ):
+            left = self.categories
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:378]: can you add a comment that the ordering of checks is for perf

File: pandas/core/dtypes/dtypes.py [hunk 3]
```diff
@@ -375,12 +375,28 @@ def __eq__(self, other: Any) -> bool:
...
+            right = other.categories
+            if not left.dtype == right.dtype:
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:380]: is_dtype_equal

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/dtypes.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:380]: thatd require a circular import (actually id like to put is_dtype_equal "above" this file, but thats for another day)

**Review by jreback — APPROVED** [review]

**Pull request merged**